### ISSUE TITLE
max-heap-size attribute cause start of 4 parallel (2x javaws.exe & 2x java.exe) vm processes which cause hangs

### ISSUE THREAD

**Issue by fischerRo** [AdoptOpenJDK/IcedTea-Web#273]
if jnlp the "j2se" tag within "ressource" tag contains an max-heap-size="256M" sample :
`<resources> 
    <j2se max-heap-size="256M" version="*" /> 
    <jar href="lib/weblauncher.jar"/>
  </resources> ` 
then the call cause to run 2 javaws process and to java process. Only one gui ist visiblity but both are working and cause hang problems.

![Processes for java webstart](https://user-images.githubusercontent.com/48080326/58895145-9797e580-86f3-11e9-9a91-9bfa85294389.png)

**Comment by judovana** [comment]
This is correct behavior. You use one VM to load jnlp, and here you find that xmx is desired. So you invoke new VM, where xmx is passed in, and that one is launching the final app.

**Comment by fischerRo** [comment]
Perhaps i have the Problem not good explain. There are 4 Processes 2 javaws.exe and 2 java.exe.  The calling ist startet over doppelklick for an jnlp File over File association see :
![grafik](https://user-images.githubusercontent.com/48080326/59091864-8677f000-8910-11e9-976b-e46d4a03050e.png)
None of the javaws.exe Prozesses is an Child-Prozess of any of the other Processes. Each java.exe Prozess ist only the child from one of the javaws.exe Prozesses. See also the Graphic in the first comment.
 I think both java.exe Prozesses are running the  weblauncher.jar  although only one gui is showing. On startup our Code in weblauncher.jar download something into the same folder so there are going into some hang-conflict because there are called twice. Without max-heap-size Attribute there is no Problem.

**Comment by maru-sama** [comment]
I am experiencing most likely the same issue when testing this with Iuclid6 (https://iuclid6.echa.europa.eu/).
When trying to open the jnlp file by double clicking it, startup hangs because the second java process is locked by the first one. If I close the first one, the program starts up normally. 
If I use the same jnlp file from the command-line two java process are started as well but for whatever reason they are NOT locking each other so the application starts.

**Comment by judovana** [comment]
>Perhaps i have the Problem not good explain. There are 4 Processes 2 javaws.exe and 2 java.exe. The calling ist startet with double-click for an jnlp File with File association see 

That is still the same. First javaws.exe, launched java (as icedtea-wb is written in java) , that java starts  another javaws.exe and that starts java vm whcich run itw jars, and process jnlp.

Interesting is what maru-sama says: 
> If I use the same jnlp file from the command-line two java process are started as well but for whatever reason they are NOT locking each other so the application starts.

So this smell like bug in windows x gui x rust implemenatiton chain only. My windows expertise is very low, I need to ask for help on this field.

**Comment by judovana** [comment]
@lherschi  very likely can help

**Comment by judovana** [comment]
Any way how to contact lherschi ?

**Comment by lherschi** [comment]
As a workaraound should first help the switch -Xnofork. Until the refactoring of the bat launchers, the switch was always set there.
The rust launcher starts the java process on windows with the process creation flag CREATE_NO_WINDOW if it is not running inside a console to prevent opening a console window. You could comment it out and test it or even use the flag DETACHED_PROCESS instead. Unfortunately, I can not reproduce the error and therefore can not test my ideas.

**Comment by maru-sama** [comment]
@lherschi Is it possible that you provide a build without CREATE_NO_WINDOW and a second one with DETACHED_PROCESS for testing?

**Comment by maru-sama** [comment]
Hello,
I just tested with -Xnofork and it fixes my issues. There is only one java process and no hangs.

**Comment by lherschi** [comment]
[icedtea-web-1.8.1.msi.zip](https://github.com/AdoptOpenJDK/IcedTea-Web/files/3280108/icedtea-web-1.8.1.msi.zip)
This is a build with flag DETACHED_PROCESS.

**Comment by lherschi** [comment]
[icedtea-web-1.8.1.msi.zip](https://github.com/AdoptOpenJDK/IcedTea-Web/files/3280140/icedtea-web-1.8.1.msi.zip)
This is a build without process creation flag.

**Comment by maru-sama** [comment]
I  did a quick test with both of them and neither of them works Starting from a command-line works fine though. The second process hangs at

java.io.IOException: Cant create file C:\Users\<user<\.config\icedtea-web\security\trusted.certs.temp
 at net.sourceforge.jnlp.util.FileUtils.createRestrictedFile(FileUtils.java:246)     at net.sourceforge.jnlp.util.FileUtils.createRestrictedFile(FileUtils.java:172)     at net.sourceforge.jnlp.security.KeyStores.createKeyStoreFromFile(KeyStores.java:352)     at net.sourceforge.jnlp.security.KeyStores.getKeyStore(KeyStores.java:141)     at net.sourceforge.jnlp.security.KeyStores.getKeyStore(KeyStores.java:119)     at net.sourceforge.jnlp.security.KeyStores.getCertKeyStores(KeyStores.java:183)     at net.sourceforge.jnlp.security.VariableX509TrustManager.<init>(VariableX509TrustManager.java:99)     at net.sourceforge.jnlp.security.VariableX509TrustManager.getInstance(VariableX509TrustManager.java:435)     at net.sourceforge.jnlp.security.VariableX509TrustManagerJDK7.<init>(VariableX509TrustManagerJDK7.java:53)     at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)     at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)     at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)     at java.lang.reflect.Constructor.newInstance(Constructor.java:423)     at net.sourceforge.jnlp.runtime.JNLPRuntime.getSSLSocketTrustManager(JNLPRuntime.java:348)     at net.sourceforge.jnlp.runtime.JNLPRuntime.initialize(JNLPRuntime.java:279)     at net.sourceforge.jnlp.runtime.Boot.init(Boot.java:349)     at net.sourceforge.jnlp.runtime.JnlpBoot.run(JnlpBoot.java:58)     at net.sourceforge.jnlp.runtime.Boot.run(Boot.java:270)     at net.sourceforge.jnlp.runtime.Boot.run(Boot.java:63)     at java.security.AccessController.doPrivileged(Native Method)     at net.sourceforge.jnlp.runtime.Boot.main(Boot.java:210) 
java.io.IOException: Cant create file C:\Users\guntsmi1\.config\icedtea-web\security\trusted.jssecerts.temp     at net.sourceforge.jnlp.util.FileUtils.createRestrictedFile(FileUtils.java:246)     at net.sourceforge.jnlp.util.FileUtils.createRestrictedFile(FileUtils.java:172)     at net.sourceforge.jnlp.security.KeyStores.createKeyStoreFromFile(KeyStores.java:352)     at net.sourceforge.jnlp.security.KeyStores.getKeyStore(KeyStores.java:141)     at net.sourceforge.jnlp.security.KeyStores.getKeyStore(KeyStores.java:119)     at net.sourceforge.jnlp.security.KeyStores.getCertKeyStores(KeyStores.java:178)     at net.sourceforge.jnlp.security.VariableX509TrustManager.<init>(VariableX509TrustManager.java:99)     at net.sourceforge.jnlp.security.VariableX509TrustManager.getInstance(VariableX509TrustManager.java:435)     at net.sourceforge.jnlp.security.VariableX509TrustManagerJDK7.<init>(VariableX509TrustManagerJDK7.java:53)     at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)     at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)     at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)     at java.lang.reflect.Constructor.newInstance(Constructor.java:423)     at net.sourceforge.jnlp.runtime.JNLPRuntime.getSSLSocketTrustManager(JNLPRuntime.java:348)     at net.sourceforge.jnlp.runtime.JNLPRuntime.initialize(JNLPRuntime.java:279)     at net.sourceforge.jnlp.runtime.Boot.init(Boot.java:349)     at net.sourceforge.jnlp.runtime.JnlpBoot.run(JnlpBoot.java:58)     at net.sourceforge.jnlp.runtime.Boot.run(Boot.java:270)     at net.sourceforge.jnlp.runtime.Boot.run(Boot.java:63)     at java.security.AccessController.doPrivileged(Native Method)     at net.sourceforge.jnlp.runtime.Boot.main(Boot.java:210) 
java.io.IOException: Cant create file C:\Users\guntsmi1\.config\icedtea-web\security\trusted.clientcerts.temp     at net.sourceforge.jnlp.util.FileUtils.createRestrictedFile(FileUtils.java:246)     at net.sourceforge.jnlp.util.FileUtils.createRestrictedFile(FileUtils.java:172)     at net.sourceforge.jnlp.security.KeyStores.createKeyStoreFromFile(KeyStores.java:352)     at net.sourceforge.jnlp.security.KeyStores.getKeyStore(KeyStores.java:141)     at net.sourceforge.jnlp.security.KeyStores.getKeyStore(KeyStores.java:119)     at net.sourceforge.jnlp.runtime.JNLPRuntime.initialize(JNLPRuntime.java:276)     at net.sourceforge.jnlp.runtime.Boot.init(Boot.java:349)     at net.sourceforge.jnlp.runtime.JnlpBoot.run(JnlpBoot.java:58)     at net.sourceforge.jnlp.runtime.Boot.run(Boot.java:270)     at net.sourceforge.jnlp.runtime.Boot.run(Boot.java:63)     at java.security.AccessController.doPrivileged(Native Method)     at net.sourceforge.jnlp.runtime.Boot.main(Boot.java:210)

since this is already open by the first jvm. Closing the first jvm allows the second one to proceed.

**Comment by maru-sama** [comment]
I just tested this from the command-line. I also get these exceptions but after that it continues with more exceptions and loads up the application fine.

**Comment by lherschi** [comment]
The path in the exception is strange. Somehow there is a missing backslash or is your windows user user.config? Did you set deployment.user.security.trusted.clientauthcerts in your deployment.properties?

**Comment by maru-sama** [comment]
I put it in a code block now so it should look better. The path is fine I just replace my username with "user". As for the deployment.properties, the complete application including the jnlp file is from a third party so I cannot answer this.

**Comment by lherschi** [comment]
deployment.properties is the config-file from icedtea-web. 
Icedtea-web does not manage to create keystore files for you. The files are first created as a .tmp file and then renamed without the .tmp at the end. But that only happens if there are no keystore files. Since this is always tried with you, icedtea-web probably has no write permissions in the .config directory

**Comment by maru-sama** [comment]
Ok, so I fixed this issue with the key files by completely removing the .config directory and it gets a little further (displays the splash-screen) but then it hangs as well. Killing the first java process makes it continue though. Attached you will find the log although I doubt that it helps much.
[broken.txt](https://github.com/AdoptOpenJDK/IcedTea-Web/files/3280937/broken.txt)

**Comment by maru-sama** [comment]
Starting the same from the cli or closing the first jvm shows the same exceptions, but it continues afterwards with

> result: http://iuclid.company.com/webstart/

then some more exceptions and then
>  result: http://iuclid.company.com/webstart/
>  result: http://iuclid.company.com/webstart/
>  result: http://iuclid.company.com/webstart/
> Selecting proxy for: http://iuclid.company.com/webstart/
> Selected proxies: [DIRECT]
> Selecting proxy for: http://echa.europa.eu
> Selected proxies: [DIRECT]
> Selecting proxy for: http://iuclid.company.com/webstart//client-config
> Selected proxies: [DIRECT]
> Selecting proxy for: http://iuclid.company.com/webstart//client-jars
> .....

**Comment by lherschi** [comment]
For the java.nio.file.InvalidPathException I made a pull request to fix this. The whole ClassNotFoundException comes from the logging framework and the groovy library. Unfortunately, this is the case of a failed Class.forName being caught cleanly in the application, as well as being logged by the JNLPClassLoader. All these exceptions probably have nothing to do with the deasdlock problem anyway.

**Comment by maru-sama** [comment]
Can i somehow check with the console what's deadlocking exactly?

**Comment by lherschi** [comment]
I do not know what to check on the console yet. Without proper debugging, you will probably not be able to find the error. Here is another version in which javaws.exe no longer distinguishes between konsole and gui. Please try again with the version again both variants. If the behavior does not change with the version, the error is more likely to be in the java runtime or in windows itself.
[icedtea-web-1.8.1.msi.zip](https://github.com/AdoptOpenJDK/IcedTea-Web/files/3289978/icedtea-web-1.8.1.msi.zip)

**Comment by maru-sama** [comment]
Hello, I can confirm that this version opens without any hangs. A command-line window pops-up though but I think this is expected since you no longer distinguish between console and gui.

**Comment by lherschi** [comment]
Kk, then we try to narrow down the error. Please test this version.
[icedtea-web-1.8.1.msi.zip](https://github.com/AdoptOpenJDK/IcedTea-Web/files/3290175/icedtea-web-1.8.1.msi.zip)

**Comment by maru-sama** [comment]
Now two command-line windows pop up, but with java instead of javaws, no output in either of them and the same hang.

**Comment by lherschi** [comment]
With which java major version do you test?

**Comment by maru-sama** [comment]
Java 8 nightly OpenJDK8U-jdk_x64_windows_hotspot_8u212b04. I installed it this one to get jps  and jstack

**Comment by lherschi** [comment]
Please try this one.
[icedtea-web-1.8.1.msi.zip](https://github.com/AdoptOpenJDK/IcedTea-Web/files/3290318/icedtea-web-1.8.1.msi.zip)

**Comment by maru-sama** [comment]
Same issue. Two windows and it hangs.

**Comment by lherschi** [comment]
Ok, I'm starting to run out of ideas. Please test this one.
[icedtea-web-1.8.1.msi.zip](https://github.com/AdoptOpenJDK/IcedTea-Web/files/3300373/icedtea-web-1.8.1.msi.zip)

**Comment by maru-sama** [comment]
Hi this one startup fine, but I see two command-line java windows without content.

**Comment by lherschi** [comment]
Yes, the console windows were still intended for the version. Here is a version without the console window.
[icedtea-web-1.8.1.msi.zip](https://github.com/AdoptOpenJDK/IcedTea-Web/files/3301700/icedtea-web-1.8.1.msi.zip)

**Comment by maru-sama** [comment]
Perfect, in total 4 processes are started but the application starts up fine. Just a question. Can't you fork the second process and then close the first one so the first jvm is closed, or isn't this possible in windows?

**Comment by lherschi** [comment]
Two Java processes are wanted here, one for Webstart itself and one for the application. If you kill the Webstart process, you probably lose the SecureContext and you can no longer access the keystore. For me, but also raises the question of why the javaws binary is called from the Webstart launcher (which then opens the second jvm), instead of calling right here the second jvm directly. But that would have to answer one of the main developers.

**Comment by sclassen** [comment]
only after parsing the JNLP file is the information available to start the second JVM with the correct parameters and arguments. Therefor we need to start ITW twice, first with default settings and after evaluating the JNLP with the settings for the specific application

**Comment by lherschi** [comment]
https://github.com/AdoptOpenJDK/IcedTea-Web/pull/293

**Comment by karianna** [comment]
@lherschi Is it possible to create a backport for the 1.8.x branch as well?

**Comment by lherschi** [comment]
https://github.com/AdoptOpenJDK/IcedTea-Web/pull/300

**Comment by maru-sama** [comment]
Hi, sorry for mentioning this here but I did not find a place to report this. The nightly builds are currently not working for the win64 target. Could someone take a look at it or tell me where I can report this?

**Comment by sclassen** [comment]
@maru-sama can you open a new issue for this pleas

**Comment by maru-sama** [comment]
@sclassen I have troubles figuring out in which respository. Is openjdk-infrastructure the correct one?

**Comment by sclassen** [comment]
Does this describe what you are seeing? #286

**Comment by maru-sama** [comment]
No, the jenkins build environemtn for the nightly win64 builds seems to have issues.
https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk8u/job/jdk8u-windows-x64-hotspot/

I just wanted to test one of the current nightlies with this commit to make sure everything works there, but can't since there is no build for it for several days.
That's why I want to know in which repo I should raise this issue, thinking that openjdk-infrastructure is the correct one.
